### PR TITLE
Rewrite the unsafe parts

### LINKED ISSUE #13: Iterator size_hint is wrong for range()
[This commit](https://github.com/b05902132/rust-skiplist/commit/f618885b3dc0f691e161ab0187b730c8de4c775f)
adds a new test that fails.

It seems size_hint() gives wrong size whenever the range includes the first element in the list.

### LINKED ISSUE #14: Aliasing a Box is undefined behavior
Well this sucks, but it seems that you may not alias a `Box`.

https://github.com/jonhoo/left-right/issues/74

I believe this means you cannot store a `Box` and a pointer to the box content. You will need to call `Box::into_raw` before putting it into a SkipNode.

If you are too busy to fix that, I'm currently planning to rewrite this library as an exercise. All unsafe code will reside in `skipnode.rs`. But I cannot guarantee the quality of my code 😅

### PR DESCRIPTION
**Pull request by b05902132**
All public methods are finished.

Fix #13 fix #14.

Addressed #10.

Need more tests for UB though.

### CODE REVIEW

**Comment by longwusha** [comment]
wonderful job

**Comment by JP-Ellis** [comment]
This is absolutely amazing!  It's putting my initial work to shame :sweat_smile:

**Comment by JP-Ellis** [comment]
Wow, this is such a huge change!

If you don't mind, I'll look over the code tonight or over weekend, but from the look of all the tests, it seems like it should be pretty  good!

**Comment by b05902132** [comment]
Sure. You should not merge without reviewing.

**Comment by JP-Ellis** [comment]
Overall it looks pretty good.  There's only minor things I can see, some I can fix myself.  There's a few questions I have (especially regarding safety) which I think need to be made really explicit with examples / counter examples to avoid issues down the line.

**Comment by b05902132** [comment]
Hi, I added some comments for the most confusing part of my code.
However I'm not sure they are enough. If there are other parts that need more comments, please tell me.

Also, can you add me as a collaborate? This way you don't have to maintain my code :)

**Review by JP-Ellis — APPROVED** [review]
Looks all good to me!  And yeah I'll add you as a collaborator.  You've done such great work.

I want to tidy up a few small things, but otherwise this is really fantastic!

File: src/skipnode.rs [hunk 1]
```diff
@@ -46,43 +53,426 @@ impl<V> SkipNode<V> {
...
+    }
+
+    pub fn next_mut(&mut self) -> Option<&mut Self> {
+        // SAFETY: all links either points to something or is null.
+        unsafe { self.links[0].as_mut().map(|p| p.as_mut()) }
+    }
+
+    /// Takes the next node and set next_node.prev as null.
+    ///
+    /// SAFETY: please make sure no link at level 1 or greater becomes dangling.
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:104]: When you delete a node in this list so it no longer exists, but a pointer in this list still point to it, that pointer is dangling.

I think adding examples will make the code more confusing. I will just say "The caller should fix the links" instead.

File: src/skipnode.rs [hunk 2]
```diff
@@ -46,43 +53,426 @@ impl<V> SkipNode<V> {
...
             links: iter::repeat(None).take(level + 1).collect(),
             links_len: iter::repeat(0).take(level + 1).collect(),
         }
     }
 
-    /// Consumes the node returning the value it contains.
+    /// Consumes the node returning the item it contains.
     pub fn into_inner(mut self) -> Option<V> {
-        self.value.take()
+        self.item.take()
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:84]: This take `self` by value, so by the end of the scope it's implicitly dropped.

File: src/skipnode.rs [hunk 3]
```diff
@@ -46,43 +53,426 @@ impl<V> SkipNode<V> {
...
+    ///
+    /// SAFETY: please make sure no link at level 1 or greater becomes dangling.
+    pub unsafe fn take_tail(&mut self) -> Option<Box<Self>> {
+        self.links[0].take().map(|p| {
+            let mut next = Box::from_raw(p.as_ptr());
+            next.prev = None;
+            self.links_len[0] = 0;
+            next
+        })
+    }
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:112]: The caller is responsible for updating it. I'll update the comments to clarify that.

File: src/skipnode.rs [hunk 4]
```diff
@@ -46,43 +53,426 @@ impl<V> SkipNode<V> {
...
+        old_next
+    }
+    // /////////////////////////////
+    // Value Manipulation
+    // /////////////////////////////
+    //
+    // Methods that care about items carried by the nodes.
+
+    #[must_use]
+    pub fn retain<F>(&mut self, mut pred: F) -> usize
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:143]: It's added in a later commit.

File: src/skipnode.rs [hunk 5]
```diff
@@ -46,43 +53,426 @@
...
+                    }
+                }
+            }
+        }
+        removed
+    }
+
+    // /////////////////////////////
+    // Pointer Manipulations
+    // /////////////////////////////
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:206]: I'll change them to be `pub (crate)`.

It's considered a good practice to split a long function into several small functions. One benefit is that it's difficult to see if a long function is correct at a glance, but if it's decomposed to many smaller parts, you can individually check those parts.

It's mentioned in [Code Complete](https://www.amazon.com/Code-Complete-Practical-Handbook-Construction/dp/0735619670). Linked list in Linux is also implemented in [this style](https://www.amazon.com/Code-Complete-Practical-Handbook-Construction/dp/0735619670).

File: src/skipnode.rs [hunk 6]
```diff
@@ -96,41 +486,581 @@
...
+// ///////////////////////////////////////////////
+
+/// A SeekListAction seeks a node on the list and do something, e.g. insertion, deletion,
+/// replacement, on it.
+///
+/// This trait provides some common operations that you need to implement for such actions,
+/// and a default implementation of list traversal logic.
+///
+/// See one of the types that implements this trait for examples.
+pub trait SkipListAction<'a, T>: Sized {
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:535]: Will fix that.

File: src/skipnode.rs [hunk 7]
```diff
@@ -96,41 +486,581 @@
...
+        if level == 0 && distance != self.0 {
+            None
+        } else {
+            self.0 -= distance;
+            Some((node, distance))
+        }
+    }
+}
+
+pub fn insertion_fixup<T>(
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs]: `insert_at` calls this function (via `IndexInserter`), so no.

File: src/skipnode.rs [hunk 8]
```diff
@@ -46,43 +53,426 @@ impl<V> SkipNode<V> {
...
+        // SAFETY: all links either points to something or is null.
+        unsafe { self.links[0].as_ref().map(|p| p.as_ref()) }
+    }
+
+    pub fn next_mut(&mut self) -> Option<&mut Self> {
+        // SAFETY: all links either points to something or is null.
+        unsafe { self.links[0].as_mut().map(|p| p.as_mut()) }
+    }
+
+    /// Takes the next node and set next_node.prev as null.
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:102]: `tail` is a slang for "everything after this node".
`take` means "The caller now owns the returned object" in Rust slang.

Can the reader see what this function does if I mention those two ideas in the comments without drawing a diagram?

File: src/skipnode.rs [hunk 9]
```diff
@@ -46,43 +53,426 @@ impl<V> SkipNode<V> {
...
+
+    /// Move to the last node reachable from this node.
+    pub fn last_mut(&mut self) -> &mut Self {
+        (0..=self.level).rev().fold(self, |node, level| {
+            node.advance_while_at_level_mut(level, |_, _| true).0
+        })
+    }
+
+    /// Try to move for the given distance, only using links at the specified level.
+    /// If it's impossible, then move as far as possible.
```
[the remark this answers is not in the document]
Comment (b05902132) [reply src/skipnode.rs:284]: Yes. Thanks, English is not my first language.

**Comment by longwusha** [comment]
After such a big rewrite, it would be a good idea to release a new version.

**Comment by JP-Ellis** [comment]
@longwusha Indeed.  I just want to have the other pull request #16 merged and then I'll release a new version.

**Pull request merged**